tests/test_adapter.py

```diff
@@ -8,109 +8,136 @@
 
 from harmony.message import Message
 from harmony_netcdf_to_zarr.__main__ import main
-from harmony_netcdf_to_zarr.adapter import NetCDFToZarrAdapter
-import harmony.util
+from harmony_netcdf_to_zarr.adapter import NetCDFToZarrAdapter, ZarrException
 
-from .util.file_creation import ROOT_METADATA_VALUES, create_full_dataset, create_large_dataset
-from .util.harmony_interaction import (MOCK_ENV, mock_message_for,
-                                       parse_callbacks)
+from .util.file_creation import (ROOT_METADATA_VALUES, create_full_dataset,
+                                 create_input_catalog, create_large_dataset)
+from .util.harmony_interaction import MOCK_ENV, mock_message
 
 logger = logging.getLogger()
 
 
 def mock_mp_fork_popen_launch(class_object, process_obj):
-    """
-    This method serves as a mocker
-        for multiprocessing.popen_fork.Popen._launch method
-    Basically it will only do the work in the parent process
-        because moto is holding all the objects in memory
-        and therefore they will be gone whenever children processes quit
+    """ `moto` is used to mock S3, however, `moto` does not work with
+        multiprocessing, because it holds objects in memory, and then the
+        objects are gone when the child processes quit. This method mocks
+        `multiprocessing.popen_fork.Popen._launch` to do the work in the
+        parent process.
+
     Parameters
     ----------
     class_object: object
         class object for multiprocessing.popen_fork.Popen
     process_obj: object
         task object to be processed by multiprocessing.popen_fork.Popen._launch
     """
+
+    def attempt_to_close(*fds):
+        """ Mimic `multiprocessing.util.close_fds`, but wrap the `os.close`
+            statement with a test for `None`, to avoid issues with tests.
+
+        """
+        for fd in fds:
+            if fd is not None:
+                os.close(fd)
+
     code = 1
     parent_r, child_w = os.pipe()
     child_r, parent_w = os.pipe()
     class_object.pid = os.fork()
     if class_object.pid == 0:
-        os._exit(0)
+        code = 0
+        os._exit(code)
     else:
         code = process_obj._bootstrap()
         os.close(child_w)
         os.close(child_r)
-        class_object.finalizer = mp_util.Finalize(class_object, mp_util.close_fds,
+        class_object.finalizer = mp_util.Finalize(class_object, attempt_to_close,
                                                   (parent_r, parent_w,))
         class_object.sentinel = parent_r
 
 
 class TestAdapter(unittest.TestCase):
-    """
-    Tests the Harmony adapter
-    """
+    """ Tests the Harmony adapter """
     def setUp(self):
         self.maxdiff = None
-        self.config = harmony.util.config(validate=False)
+        self.config = harmony_config(validate=False)
+        self.metadata_dir = mkdtemp()
+
+    def tearDown(self):
+        rmtree(self.metadata_dir)
 
+    @patch('harmony_netcdf_to_zarr.adapter.download_granules')
```

Comment:
I'm a little bit disappointed to have to mock the download of the input granule as that makes this (and the large file conversion) less of an end-to-end test. However, the mock used to make the `multiprocessing` module work with `moto` was then causing issues with the `multiprocessing.Queue` used to download the input granules.

I spent quite a while looking into this - there was some odd behaviour, indeed! Mocking the download in these two tests was the simplest way I could find to make the tests work in the short term.